File: 2018_MotionProfiling/src/Polynomial.java
public class Polynomial {

    private double[] coefficients;

    public Polynomial(double... coefficients) {
        this.coefficients = coefficients;
    }

    public double evaluate(double parameter) {
        double result = 0.0;
        for (int i = 0; i < coefficients.length; i++) {
            result += Math.pow(parameter, i) * this.coefficients[i];
        }

        return result;
    }

    public double[] getCoefficients() {
        return this.coefficients;
    }

    public String toString() {
        return this.toString("x");
    }

    public String toString(String parameter) {
        if (this.coefficients.length == 0) {
            return "0";
        }

        String stringPolynomial = "";
        for (int i = 0; i < this.coefficients.length; i++) {
            if (Math.abs(coefficients[i]) > 0.00000000000001) {
                stringPolynomial += this.getStringFor(parameter, this.coefficients[i], i) + " + ";
            }
        }

        return stringPolynomial.length() == 0 ? "0" : stringPolynomial.substring(0, stringPolynomial.length() - 3);
    }

    private String getStringFor(String parameter, double coefficient, int exponent) {
        String coefficientString = "" + String.format("%.30f", coefficient);

        switch (exponent) {
            case 0:
                return coefficientString;
            case 1:
                return coefficientString + parameter;
            default:
                return coefficientString + parameter + "^" + exponent;
        }
    }

    public static Polynomial differentiate(Polynomial polynomial) {
        double[] coefficients = polynomial.getCoefficients();

        if (coefficients.length == 0) {
            return new Polynomial();
        }

        double[] derivativeCoefficients = new double[coefficients.length - 1];

        for (int i = 1; i < coefficients.length; i++) {
            derivativeCoefficients[i - 1] = coefficients[i] * i;
        }
        return new Polynomial(derivativeCoefficients);
    }

    public static Polynomial integrate(Polynomial polynomial, double constant) {
        double[] coefficients = polynomial.getCoefficients();
        double[] integralCoefficients = new double[coefficients.length + 1];

        integralCoefficients[0] = constant;
        for (int i = 0; i < coefficients.length; i++) {
            integralCoefficients[i + 1] = coefficients[i] / (double) (i + 1);
        }
        return new Polynomial(integralCoefficients);
    }

    public static Polynomial sum(Polynomial... polynomials) {
        int degree = 0;
        for (Polynomial polynomial : polynomials) {
            if (polynomial.getCoefficients().length > degree) {
                degree = polynomial.getCoefficients().length;
            }
        }

        double[] newCoefficients = new double[degree];

        for (Polynomial polynomial : polynomials) {

            double[] currentCoefficients = polynomial.getCoefficients();
            for (int i = 0; i < currentCoefficients.length; i++) {
                newCoefficients[i] += currentCoefficients[i];
            }
        }

        return new Polynomial(newCoefficients);
    }

    public static Polynomial multiply(Polynomial polynomial, double multiplier) {
        double[] coefficients = polynomial.getCoefficients();
        double[] newCoefficients = new double[coefficients.length];
        for (int i = 0; i < coefficients.length; i++) {
            newCoefficients[i] = coefficients[i] * multiplier;
        }

        return new Polynomial(newCoefficients);
    }

    public static void main(String[] args) {
        Polynomial p1 = new Polynomial(1.0, 5.0, 6.0, 3.0);
        Polynomial p2 = new Polynomial(4.0, 0.0, -1.0, 2.0);

        System.out.println(Polynomial.sum(p1, p2).toString());
        System.out.println(Polynomial.multiply(p1, 2));
    }

}

File: TrajectoryGeneration/trajectory/Polynomial.go
package trajectory

import (
	"fmt"
	"math"
)

type Polynomial struct {
	coefficients []float64
}

func NewPolynomial(coefficients ...float64) Polynomial {
	return Polynomial{coefficients}
}

func (polynomial *Polynomial) Evaluate(parameter float64) float64 {
	result := 0.0
	for i := 0; i < len(polynomial.coefficients); i++ {
		result += math.Pow(parameter, float64(i)) * polynomial.coefficients[i]
	}

	return result
}

func (polynomial Polynomial) String() string {
	return polynomial.StringForParameter("X")
}

func (polynomial Polynomial) StringForParameter(parameter string) string {
	if len(polynomial.coefficients) == 0 {
		return "0"
	}

	stringPolynomial := ""
	for index, coefficient := range polynomial.coefficients {
		if math.Abs(polynomial.coefficients[index]) > 0.00000000000001 {
			stringPolynomial += stringForTerm(parameter, coefficient, index) + "+"
		}
	}

	if len(stringPolynomial) == 0 {
		return "0"
	} else {
		return string([]rune(stringPolynomial)[:len(stringPolynomial)-1])
	}
}

func stringForTerm(parameter string, coefficient float64, exponent int) string {
	coefficientString := fmt.Sprintf("%.4f", coefficient)

	switch exponent {
	case 0:
		return coefficientString
	case 1:
		return coefficientString + parameter
	default:
		return coefficientString + parameter + "^" + fmt.Sprintf("%d", exponent)
	}
}

func Differentiate(polynomial Polynomial) Polynomial {
	if len(polynomial.coefficients) == 0 {
		return Polynomial{}
	}

	derivativeCoefficients := make([]float64, len(polynomial.coefficients)-1)
	for i := 1; i < len(polynomial.coefficients); i++ {
		derivativeCoefficients[i-1] = polynomial.coefficients[i] * float64(i)
	}

	return Polynomial{derivativeCoefficients}
}

func Integrate(polynomial Polynomial, constant float64) Polynomial {
	integralCoefficients := make([]float64, len(polynomial.coefficients)+1)
	integralCoefficients[0] = constant
	for i := 0; i < len(polynomial.coefficients); i++ {
		integralCoefficients[i+1] = polynomial.coefficients[i] / float64(i+1)
	}

	return Polynomial{integralCoefficients}
}

func Sum(polynomials ...Polynomial) Polynomial {
	degree := 0
	for _, polynomial := range polynomials {
		if len(polynomial.coefficients) > degree {
			degree = len(polynomial.coefficients)
		}
	}

	summedCoefficients := make([]float64, degree)
	for _, polynomial := range polynomials {
		for index, coefficient := range polynomial.coefficients {
			summedCoefficients[index] += coefficient
		}
	}

	return Polynomial{summedCoefficients}
}

func Multiply(polynomial Polynomial, multiplier float64) Polynomial {
	multipliedCoefficients := make([]float64, len(polynomial.coefficients))
	for index, coefficient := range polynomial.coefficients {
		multipliedCoefficients[index] = coefficient * multiplier
	}

	return Polynomial{multipliedCoefficients}
}
